src/codegen/transformations.lisp

```diff
@@ -246,3 +223,171 @@
+             (make-node-let
+              :type (node-type node)
+              :bindings (loop :for (name . node) :in (node-let-bindings node)
+                              :collect (cons name (traverse node funs new-args)))
+              :subexpr (traverse (node-let-subexpr node) funs new-args)))))
+   (cons :traverse-match
+         (lambda (node funs &key bound-variables)
+           (declare (type node-match node)
+                    (type list funs)
+                    (type util:symbol-list bound-variables)
+                    (values node-match &optional))
+           (make-node-match
+            :type (node-type node)
+            :expr (traverse (node-match-expr node) funs (list :bound-variables bound-variables))
+            :branches (mapcar
+                       (lambda (branch)
+                         (make-match-branch
+                          :pattern (match-branch-pattern branch)
+                          :body (traverse
+                                 (match-branch-body branch)
+                                 funs
+                                 (list :bound-variables
+                                       (append (pattern-variables (match-branch-pattern branch))
+                                               bound-variables)))))
+                       (node-match-branches node)))))
+   (cons :traverse-dynamic-extent
+         (lambda (node funs &key bound-variables)
+           (declare (type node-dynamic-extent node)
+                    (type list funs)
+                    (type util:symbol-list bound-variables)
+                    (values node-dynamic-extent &optional))
+           (let ((new-args (list :bound-variables
+                                 (cons (node-dynamic-extent-name node) bound-variables))))
+             (make-node-dynamic-extent
+              :type (node-type node)
+              :name (node-dynamic-extent-name node)
+              :node (traverse (node-dynamic-extent-node node) funs new-args)
+              :body (traverse (node-dynamic-extent-body node) funs new-args)))))
+   (cons :traverse-bind
+         (lambda (node funs &key bound-variables)
+           (declare (type node-bind node)
+                    (type list funs)
+                    (type util:symbol-list bound-variables)
+                    (values node-bind &optional))
+           (let ((new-args (list :bound-variables
+                                 (cons (node-bind-name node) bound-variables))))
+             (make-node-bind
+              :type (node-type node)
+              :name (node-bind-name node)
+              :expr (traverse (node-bind-expr node) funs new-args)
+              :body (traverse (node-bind-body node) funs new-args)))))))
+
+(defun traverse-with-binding-list (node funs)
+  (declare (type node node)
+           (type list funs)
+           (values node &optional))
+  (traverse node
+            (append funs binding-list-traversals)
+            (list :bound-variables nil)))
+
```

Comment:
Is there a nicer way to specify these `bound-variables`-specific traversal functions?